quinn/src/runtime.rs

```diff
@@ -75,72 +76,98 @@ pub trait AsyncUdpSocket: Send + Sync + Debug + 'static {
     fn poll_send(
         self: Pin<&mut Self>,
         transmit: &Transmit,
         cx: &mut Context,
     ) -> Poll<io::Result<()>>;
 
-    /// Maximum number of datagrams that a [`Transmit`] may encode
+    /// Maximum number of datagrams that a [`Transmit`] may encode.
     fn max_transmit_segments(&self) -> usize {
         1
     }
 
-    /// TODO(matheus23): Docs
-    /// Last ditch/best effort of sending a transmit.
-    /// Used by the endpoint for resets / close frames when dropped, etc.
+    /// Try to send a UDP datagram, if the socket happens to be write-ready.
+    ///
+    /// This may fail with [`io::ErrorKind::WouldBlock`], if the socket is currently full.
+    ///
+    /// The quinn endpoint uses this function when sending
+    ///
+    /// - A version negotiation response due to an unknown version
+    /// - A `CLOSE` due to a malformed or unwanted connection attempt
+    /// - A stateless reset due to an unrecognized connection
+    /// - A `Retry` packet due to a connection attempt when `use_retry` is set
+    ///
+    /// If sending in these cases fails, a well-behaved peer will re-try. Thus it's fine
+    /// if we drop datagrams sometimes with this function.
     fn try_send(self: Pin<&mut Self>, transmit: &Transmit) -> io::Result<()>;
 }
 
 pin_project_lite::pin_project! {
-    pub struct UdpSenderHelper<Socket, MakeFut, Fut> {
+    /// A helper for constructing [`UdpSender`]s from an underlying `Socket` type.
+    ///
+    /// This struct implements [`UdpSender`] if `MakeWritableFn` produces a `WritableFut`.
+    ///
+    /// Also serves as a trick, since `WritableFut` doesn't need to be a named future,
+    /// it can be an anonymous async block, as long as `MakeWritableFn` produces that
+    /// anonymous async block type.
+    ///
+    /// The `UdpSenderHelper` generic type parameters don't need to named, as it will be
+    /// used in its dyn-compatible form as a `Pin<Box<dyn UdpSender>>`.
+    pub struct UdpSenderHelper<Socket, MakeWritableFutFn, WritableFut> {
         socket: Socket,
-        make_fut: MakeFut,
+        make_writable_fut_fn: MakeWritableFutFn,
         #[pin]
-        fut: Option<Fut>,
+        writable_fut: Option<WritableFut>,
     }
 }
 
-impl<Socket, MakeFut, Fut> Debug for UdpSenderHelper<Socket, MakeFut, Fut> {
+impl<Socket, MakeWritableFutFn, WritableFut> Debug
+    for UdpSenderHelper<Socket, MakeWritableFutFn, WritableFut>
+{
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         f.write_str("UdpSender")
     }
 }
 
-impl<Socket, MakeFut, Fut> UdpSenderHelper<Socket, MakeFut, Fut> {
-    pub fn new(inner: Socket, make_fut: MakeFut) -> Self {
+impl<Socket, MakeWritableFutFn, WriteableFut>
+    UdpSenderHelper<Socket, MakeWritableFutFn, WriteableFut>
+{
+    /// Create a [`UdpSender`] from a socket and a function that takes
+    /// the socket's reference and produces a future that resolves to
+    /// `std::io::Result<()>`, once the socket is write-ready.
```

Comment:
A bit of a long sentence for the leader.  Since this isn't an internal comment (at least once upstreamed) maybe let's re-word this a bit to fit into a single-line summary.